File: elixir/bracket-push/bracket_push.exs
defmodule BracketPush do
  @doc """
  Checks that all the brackets and braces in the string are matched correctly, and nested correctly
  """
  @spec check_brackets(String.t) :: boolean
  def check_brackets(str) do
    str
    |> String.replace(~r/[^(\{\[)\}\]]/, "")  # strip out chars that aren't brackets
    |> String.codepoints                     
    |> Enum.reduce([], &bracket_stack/2)      # expect to be reduced to empty list
    |> Enum.empty?
  end


  defp bracket_stack("[", acc), do: ["]" | acc]     # push brackets onto stack
  defp bracket_stack("(", acc), do: [")" | acc]     
  defp bracket_stack("{", acc), do: ["}" | acc]     
  defp bracket_stack(x, [a|as]) when x == a, do: as # match bracket, pop stack
  defp bracket_stack(_x, _acc), do: [false]         # Closing bracket no opening ... ruin stack

end

File: ocaml/bracket-push/bracket_push.ml

let brackets = [
  ('(', ')'); 
  ('[', ']'); 
  ('{', '}');
  ]

let is_bracket f chr= brackets |> List.map f |> List.exists ((=) chr)
let is_opening_bracket = is_bracket fst
let is_closing_bracket = is_bracket snd

let matched_pair opening closing = 
   brackets |> Base.List.find ~f: (fun x -> snd x = closing) |> Base.Option.map ~f:fst = Some opening

let pop_nested_pair stack closing = match stack with 
  | (hd::tl) when matched_pair hd closing -> tl
  | _ -> closing :: stack

let bracket_stack stack chr = match (stack, chr) with
  | (stack, chr) when is_opening_bracket chr -> chr :: stack
  | (stack, chr) when is_closing_bracket chr -> pop_nested_pair stack chr
  | _ -> stack

let are_balanced str = str |> Base.String.to_list |> List.fold_left bracket_stack [] = [];;